### ISSUE TITLE
Forge LLMs like Qwen3-8B decode tok/s degrades with output length (though online server)

### ISSUE THREAD

**Issue by kmabeeTT** [tenstorrent/tt-inference-server#4278]
## Summary
On the tt-inference-server online server, Qwen3-8B per-request decode throughput **falls as the requested output length (OSL) grows** (concurrency 1, fixed input). The decode rate should be roughly constant with OSL. Evidence below localizes the cause to the **serving path** (`AsyncLLMEngine.generate()` per-token output handling), not the engine/device — the tt-xla vLLM engine decodes at a flat rate under the same config.

## Background (how this surfaced)
This came out of profiling unexpectedly long eval runtimes. A 2-sample eval (the 2 samples run concurrently, single layer, each generating ~12k output tokens) took ~30 min. At the expected ~40 tok/s/sample that should have been ~5 min; the observed effective rate was only ~6.7 tok/s/sample (~12k tokens / ~1800 s) — roughly 6× slower — which prompted this decode-rate-vs-OSL investigation.

## Observation (concurrency 1, ISL 128)

Online, single layer (NUM_HIDDEN_LAYERS=1):

| OSL | 127 | 256 | 1022 | 2046 | 4083 | 8192 | 16373 |
|-----|----|----|-----|-----|-----|-----|------|
| tok/s/req | 47.9 | 47.4 | 42.8 | 38.4 | 31.7 | 23.8 | 16.0 |

Online, full model (still degrades, less steeply):

| OSL | 128 | 255 | 1021 | 1973 | 3863 | 7661 |
|-----|----|----|-----|-----|-----|-----|
| tok/s/req | 14.1 | 13.9 | 13.5 | 12.7 | 11.8 | 10.4 |

Sweeping **input** length (ISL) on the same online server does **not** reproduce the slowdown — we expected it might, but it doesn't. tok/s/req stays ~46 even at ISL 16384; only TTFT/prefill grows and agg tok/s falls (prefill amortized over a fixed 128-token output). So decode rate depends on tokens *generated* (OSL), not total context depth (OSL fixed at 128 here):

| ISL | 128 | 1024 | 2048 | 4096 | 8192 | 16384 |
|-----|----|-----|-----|-----|-----|------|
| tok/s/req | 46.3 | 46.8 | 47.1 | 46.0 | 46.4 | 45.8 |
| TTFT ms | 68 | 421 | 903 | 1878 | 4110 | 9830 |

## Evidence it's the serving path, not the engine/device
- **Offline tt-xla vLLM benchmark, single layer, configured to match the server** (max_num_seqs=32, max_model_len=40960, min_context_len=128, same additional_config, single request): decode is **flat** across OSL — the engine produces tokens at a constant rate regardless of OSL:

  | OSL | 128 | 1024 | 4096 |
  |-----|----|-----|-----|
  | decode tok/s | 82.8 | 82.1 | 79.9 |
- **stream=true vs stream=false** (online): both degrade identically (≈49→33 over OSL 128→4096) → not SSE/streaming transport.
- **USE_DYNAMIC_BATCHER=false** (online): still degrades (≈49→33) → not the dynamic batcher.
- **ttnn IR** (online vs offline, via extract_mlir_graphs.py): same op set/structure.
- Remaining suspect: `AsyncLLMEngine.generate()` per-token async output processing (per-step RequestOutput construction + `RequestOutputKind.DELTA` detokenization), which the offline synchronous `LLM.generate()` path does not use. The streaming loop sends DELTA chunks (O(1)/token), so it is **not** a cumulative re-detokenize bug.
- Aside: a smaller single-layer offline config (batch 1, lower max_model_len) ran at ~120 tok/s — noted only for context; not the comparison used above.

## Two distinct serving-path costs
Comparing the matched offline control (batch-32, flat ~82 tok/s) against the online server isolates two separate problems, both in the serving path:
- **Fixed offset** — at OSL=128 (before any degradation), the same single-layer config drops from **~82 tok/s (tt-xla offline) to ~46 tok/s (tt-inference-server online)**: a ~1.7× per-token serving cost that is independent of output length.
- **Scaling** — on top of that fixed offset, online decode *further* degrades as OSL grows (~46 → 16 tok/s over OSL 128 → 16K). This is the primary issue.

## Repro
- **tt-inference-server** (online, degrades) — branch `kmabee/forge_llm_chunked_prefill.debug` (launch fix commit `cec043795`): run `tt-media-server/launch_qwen3_8b.sh` with `NUM_HIDDEN_LAYERS=1`, then `~/scripts/test_all_llm_servers.sh --concurrent 1 --isl 128 --osl <N>`.
- **tt-xla** (offline control, flat) — branch `kmabee/chunked_prefill_isue_4986_explore.rebase` (commit `ddcb7277d`): run `tests/benchmark/test_vllm_benchmarks.py::test_vllm_qwen3_8b_decode_sweep_match_online` (+ `tests/benchmark/sweep_qwen3_8b_decode.sh`).

## Recommended next step (name the exact cost)
Log the engine-internal decode rate in the runner's `async for` loop from `RequestOutput.metrics` (`last_token_ts - first_token_ts`). If engine-internal is flat while the client sees the degradation, the cost is entirely the Python async per-token output handling. (py-spy would pinpoint the call but needs sudo here: ptrace_scope=1.)

--

Debug ongoing within tt-inference-server, no need to assign to perf team yet.

**Comment by kmabeeTT** [comment]
### Root cause: repetition-penalty token counts are rebuilt O(N) per decode step (O(N²) per request) on the host, triggered by a default `repetition_penalty=1.1`

The decode slowdown is **host-side penalty preprocessing** in the tt-xla vLLM plugin — not vLLM, the async engine, the device kernel, or the serving layer.

**Code** — `tt-xla integrations/vllm_plugin/vllm_tt/metadata.py::_compute_token_counts`, run every decode step whenever a penalty is active:

```python
counts = torch.zeros(padded_num_reqs, vocab_size, dtype=torch.float32)  # [32, 151936] ≈ 19 MB, fresh every step
for i, token_ids in enumerate(req_output_token_ids[:padded_num_reqs]):
    if token_ids:
        for token_id in token_ids:        # O(output_length) Python loop, every decode step
            counts[i, token_id] += 1
```

It re-counts the **entire generated history** into a fresh `[batch, vocab]` CPU tensor every step → O(N) per step, **O(N²)** per request. This is host CPU work *outside* the compiled graph; the on-device penalty math is fixed-shape (`[batch, vocab]`) and constant. So identical IR and flat per-step *device* time are both fully consistent with the growth — it's host glue, not the graph. (Measured directly: this function alone grows ~1.8 ms → 5.5 ms from 191 → 791 output tokens, ~0.006 ms/token, accounting for the whole per-step wall-time growth.)

**Trigger:** tt-media-server defaults `repetition_penalty=1.1` for every request (`tt_model_runners/vllm_runner.py::SAMPLING_DEFAULTS` + `config/constants._DEFAULT_SAMPLING_PARAMS`). With `repetition_penalty=1.0` the penalty path is skipped (`no_penalties=True`) — which is why the offline benchmark and a default standalone run were flat.

**Reproduced in pure tt-xla** (no tt-media-server, single layer, `AsyncLLMEngine`, one request, only variable = `repetition_penalty`):

| OSL | rep=1.0 ms/step | rep=1.1 ms/step | rep=1.1 tok/s | online server tok/s |
|---|---|---|---|---|
| 128 | 12.3 (flat) | 22.2 | 44.8 | ~48 |
| 4096 | 12.9 (flat) | 22 → 43 | 30.1 | ~31 |

rep=1.0 is dead flat across the whole request; rep=1.1 grows linearly and matches the server. This explains both observed costs: the ~1.7× baseline offset (penalty preprocessing) and the O(N²) growth. Repro: `tt-xla tests/benchmark/async_llm_decode_repro.py` (`REP_PENALTY` env).

**Fix:**
1. **tt-inference-server (immediate):** stop defaulting `repetition_penalty=1.1` (use `1.0` / opt-in). Restores ~80 tok/s single-layer decode and fixes the ~6× slow eval runs.
2. **tt-xla (proper):** make `_compute_token_counts` / `_compute_prompt_mask` incremental — maintain a persistent `[batch, vocab]` count and add the single new token per step (O(1)) instead of re-scanning history and re-allocating ~19 MB each step. Same semantics (repetition penalty *correctly* covers all prior tokens — that's the standard definition, not a sliding window), just not recomputed from scratch.

- *Note:* this took a long process of elimination — vLLM `AsyncLLMEngine`, the device/decode kernel, async scheduling, the sampling kernel, prefix caching, trace replay, GIL/busy-loop contention, and output-queue backpressure were each measured and ruled out before the penalty default surfaced.

**Comment by kmabeeTT** [comment]
Datapoints for each of the ~Fix options in the previous comment, again using the same single-layer Qwen3-8B server used in prior experiments.

**1. Explicitly use `repetition_penalty=1.0` on the request (or change the runner default from `1.1`).** Original sweep is flat again (and jumps from ~46 → ~76 tok/sec) — avoids the issue as expected:

```
========================================================================
SWEEP SUMMARY  (agg tok/s = total output tokens / wall;  tok/s/req = mean per-stream decode)
========================================================================
model      conc   ISL     OSL   TTFT ms   tok/s/req   agg tok/s   wall s
--------   ----   ---   -----   -------   ---------   ---------   ------
Qwen3-8B      1   128     128        76        76.8        74.0     1.73
Qwen3-8B      1   128     255        59        76.7        75.7     3.37
Qwen3-8B      1   128    1024        60        76.7        76.4    13.40
Qwen3-8B      1   128    2045        59        76.0        75.9    26.95
Qwen3-8B      1   128    4094        58        75.0        74.9    54.67
Qwen3-8B      1   128    8191        58        72.6        72.5   112.91
Qwen3-8B      1   128   16382        62        69.3        69.3   236.34
```

**2. Alternatively, with a candidate fix in tt-xla to make `repetition_penalty` efficient** (described in the previous comment — keep a persistent `[batch, vocab]` output-token-count tensor and fold only the newly-generated token each step, instead of re-scanning the full output history → O(1)/step instead of O(N)/step), we can run with `repetition_penalty=1.1` and the O(N²) cost is gone — flat across OSL:

```
========================================================================
SWEEP SUMMARY  (agg tok/s = total output tokens / wall;  tok/s/req = mean per-stream decode)
========================================================================
model      conc   ISL     OSL   TTFT ms   tok/s/req   agg tok/s   wall s
--------   ----   ---   -----   -------   ---------   ---------   ------
Qwen3-8B      1   128     128        66        51.1        50.2     2.55
Qwen3-8B      1   128     254        65        50.4        49.9     5.09
Qwen3-8B      1   128    1024        65        50.8        50.6    20.22
Qwen3-8B      1   128    2043        64        50.1        50.0    40.85
Qwen3-8B      1   128    4091        65        50.1        50.1    81.67
Qwen3-8B      1   128    8178        66        49.5        49.5   165.37
Qwen3-8B      1   128   16377        65        48.9        48.9   335.07
```

These two are complementary, and the plan is to do both:
- **(a)** change the Forge vLLM runner default to `repetition_penalty=1.0` — it's currently the only runner defaulting to `1.1`; the sibling Forge TP runners (Llama-70B, Qwen-32B, Gemma-4-31B), the metal runner, and the repo-wide default all already use `1.0`; and
- **(b)** land the tt-xla fix so opting into `1.1` per-request stays cheap (flat ~50 vs ~76 tok/s, no decay).

---

**End-to-end through evals (lm-eval `r1_gpqa_diamond`, 2 samples, batch 2, ~12288 tok/req).** The fix holds on the serving path too. Progression below; the last row is a **preview** that includes a *separate* seed fix (see note) and lands at **~10× over the original baseline**:

| eval run | wall | per-req tok/s | ms/step | vs baseline |
|---|---|---|---|---|
| baseline — `rep=1.1` + seed (as evals ran it) | 1814.5 s | 6.8 | 147.6 | 1.0× |
| #4278 addressed — `rep=1.0` (still seeded) | 877.5 s | 14.0 | 71.4 | 2.1× |
| **+ seed dropped (separate issue) — `rep=1.0`** | **170.5 s** | **72.1** | **13.9** | **≈10.6×** |

(The tt-xla fix keeps `rep=1.1` flat end-to-end too: `rep=1.1` + seed-dropped runs at 50.0 tok/s / 20.0 ms/step — i.e. the penalty becomes a flat ~6 ms/step constant instead of an O(N²) that averaged ~75 ms/step at 12k tokens.)

**Note — separate issue, not #4278:** while validating through evals I found that lm-eval always sends a per-request `seed`, which on the TT device sampler triggers a *separate* slow path (per-step CPU noise build + ~19 MB host→device transfer, ~5× decode) and isn't even honored on-device (tt-xla#4539). Dropping/ignoring the seed is what takes the eval from 2.1× to ~10.6×. This is independent of the `repetition_penalty` O(N²) issue tracked here and will be filed as its own issue.
